package net.runelite.client.plugins.projectx.ourania;

import java.time.Duration;
import java.time.Instant;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;
import javax.inject.Inject;
import net.runelite.api.Constants;
import net.runelite.api.GameState;
import net.runelite.api.coords.WorldPoint;
import net.runelite.api.gameval.ItemID;
import net.runelite.api.gameval.NpcID;
import net.runelite.api.gameval.ObjectID;
import net.runelite.client.plugins.gpu.GpuPlugin;
import net.runelite.client.plugins.projectx.ProjectX;
import net.runelite.client.plugins.projectx.Script;
import net.runelite.client.plugins.projectx.breakhandler.BreakHandlerScript;
import net.runelite.client.plugins.projectx.pouch.Pouch;
import net.runelite.client.plugins.projectx.ourania.enums.OuraniaState;
import net.runelite.client.plugins.projectx.ourania.enums.Path;
import net.runelite.client.plugins.projectx.util.antiban.Rs2Antiban;
import net.runelite.client.plugins.projectx.util.antiban.enums.Activity;
import net.runelite.client.plugins.projectx.util.bank.Rs2Bank;
import net.runelite.client.plugins.projectx.util.camera.Rs2Camera;
import net.runelite.client.plugins.projectx.util.inventory.Rs2Inventory;
import net.runelite.client.plugins.projectx.util.inventory.Rs2ItemModel;
import net.runelite.client.plugins.projectx.util.inventory.RunePouchType;
import net.runelite.client.plugins.projectx.util.magic.Rs2Magic;
import net.runelite.client.plugins.projectx.util.magic.Rs2Spellbook;
import net.runelite.client.plugins.projectx.util.magic.Rs2Spells;
import net.runelite.client.plugins.projectx.util.math.Rs2Random;
import net.runelite.client.plugins.projectx.util.misc.Rs2Potion;
import net.runelite.client.plugins.projectx.api.npc.models.Rs2NpcModel;
import net.runelite.client.plugins.projectx.util.player.Rs2Player;
import net.runelite.client.plugins.projectx.util.walker.Rs2Walker;
import net.runelite.client.plugins.skillcalculator.skills.MagicAction;

public class OuraniaScript extends Script
{

	public static OuraniaState state;
	private final OuraniaConfig config;
	private final List<Integer> massWorlds = List.of(327, 480);
	private final OuraniaPlugin plugin;
	private int selectedWorld = 0;
    public Instant startTime;

    /**
     * Get the total runtime of the script
     *
     * @return the total runtime of the script
     */
    public Duration getRunTime() {
        if (startTime == null) return Duration.ofSeconds(0);
        return Duration.between(startTime, Instant.now());
    }

	@Inject
	public OuraniaScript(OuraniaPlugin plugin, OuraniaConfig config)
	{
		this.plugin = plugin;
		this.config = config;
	}

	@Override
	public void shutdown()
	{
		Rs2Antiban.resetAntibanSettings();
		super.shutdown();
	}

	public boolean run()
	{
        startTime = Instant.now();
		ProjectX.enableAutoRunOn = false;
		Rs2Antiban.resetAntibanSettings();
		Rs2Antiban.antibanSetupTemplates.applyRunecraftingSetup();
		Rs2Antiban.setActivity(Activity.CRAFTING_RUNES_AT_OURANIA_ALTAR);
		mainScheduledFuture = scheduledExecutorService.scheduleWithFixedDelay(() -> {
			try
			{
				if (!ProjectX.isLoggedIn())
				{
					return;
				}
				if (!super.run())
				{
					return;
				}
				long startTime = System.currentTimeMillis();

				if (!Rs2Magic.isSpellbook(Rs2Spellbook.LUNAR))
				{
					ProjectX.showMessage("Not currently on Lunar Spellbook");
					ProjectX.stopPlugin(plugin);
					return;
				}

				if (Rs2Inventory.anyPouchUnknown())
				{
					Rs2Inventory.checkPouches();
					return;
				}

				if (config.useMassWorld() && !isOnMassWorld())
				{
					if (selectedWorld == 0)
					{
						selectedWorld = massWorlds.get(Rs2Random.between(0, massWorlds.size()));
					}
					ProjectX.hopToWorld(selectedWorld);
					sleepUntil(() -> ProjectX.getClient().getGameState() == GameState.LOGGED_IN);
					return;
				}

				if (selectedWorld != 0)
				{
					selectedWorld = 0;
				}

				if (hasStateChanged())
				{
					state = updateState();
				}

				if (state == null)
				{
					ProjectX.showMessage("Unable to evaluate state");
					ProjectX.stopPlugin(plugin);
					return;
				}

				switch (state)
				{
					case CRAFTING:
						if (!Rs2Inventory.hasItem(config.essence().getItemId()) && Rs2Inventory.hasAnyPouch() && !Rs2Inventory.allPouchesEmpty())
						{
							Rs2Inventory.emptyPouches();
							return;
						}
						ProjectX.getRs2TileObjectCache().query().withId(ObjectID.RC_ZMI_DUNGEON_CRACKED_CENTER_ALTAR).interact("craft-rune");
						Rs2Inventory.waitForInventoryChanges(5000);
						break;
					case RESETTING:
						if (Rs2Player.getWorldLocation().distanceTo(new WorldPoint(2468, 3246, 0)) > 24)
						{
							Rs2Magic.cast(MagicAction.OURANIA_TELEPORT);
						}
						sleepUntil(() -> Rs2Player.getWorldLocation().distanceTo(new WorldPoint(2468, 3246, 0)) < 24);

						if (plugin.isBreakHandlerEnabled())
						{
							BreakHandlerScript.setLockState(false);
						}

						if (Rs2Inventory.hasDegradedPouch() && Rs2Magic.hasRequiredRunes(Rs2Spells.NPC_CONTACT))
						{
							Rs2Magic.repairPouchesWithLunar();
							return;
						}

						if (config.directInteract() && ProjectX.isPluginEnabled(GpuPlugin.class))
						{
							ProjectX.getRs2TileObjectCache().query().withId(ObjectID.RC_ZMI_DUNGEON_ENTRANCE).interact("Climb");
							sleepUntil(this::isNearEniola, 20000);
						}
						else
						{
							Rs2Walker.walkTo(new WorldPoint(3014, 5625, 0));
						}
						break;
					case BANKING:
						if (plugin.isRanOutOfAutoPay())
						{
							ProjectX.showMessage("You have ran out of auto-pay runes, check runepouch!");
							ProjectX.stopPlugin(plugin);
							return;
						}

						if (!Rs2Bank.isOpen())
						{
							Rs2NpcModel eniola = ProjectX.getRs2NpcCache().query().withId(NpcID.RC_ZMI_BANKER).nearest();
							if (eniola == null)
							{
								return;
							}
							eniola.click("bank");
							sleepUntil(Rs2Bank::isOpen, 3000);
							return;
						}

						if (!config.toggleProfitCalculator())
						{
							plugin.calcuateProfit();
						}

						boolean hasRunes = Rs2Inventory.items().anyMatch(item -> item.getName().toLowerCase().contains("rune") && !item.getName().toLowerCase().contains("rune pouch"));

						if (hasRunes)
						{
							if (config.useDepositAll())
							{
								Rs2Bank.depositAll();
							}
							else
							{
								// Get all RunePouchType IDs
								Integer[] runePouchIds = Arrays.stream(RunePouchType.values())
									.map(RunePouchType::getItemId)
									.toArray(Integer[]::new);

								// Get all eligible pouch IDs based on Runecrafting level
								Integer[] eligiblePouchIds = Arrays.stream(Pouch.values())
									.filter(Pouch::hasRequiredRunecraftingLevel)
									.flatMap(pouch -> Arrays.stream(pouch.getItemIds()).boxed())
									.toArray(Integer[]::new);

								// Combine RunePouchType IDs and eligible pouch IDs into a single array
								Integer[] excludedIds = Stream.concat(Arrays.stream(runePouchIds), Arrays.stream(eligiblePouchIds))
									.toArray(Integer[]::new);

								Rs2Bank.depositAllExcept(excludedIds);
								Rs2Inventory.waitForInventoryChanges(1800);
							}
						}

						if (config.useEnergyRestorePotions() && Rs2Player.getRunEnergy() <= config.drinkAtPercent())
						{
							boolean hasStaminaPotion = Rs2Bank.hasItem(Rs2Potion.getStaminaPotion());
							boolean hasEnergyRestorePotion = Rs2Bank.hasItem(Rs2Potion.getRestoreEnergyPotionsVariants());

							if ((Rs2Player.hasStaminaBuffActive() && hasEnergyRestorePotion) || (!hasStaminaPotion && hasEnergyRestorePotion))
							{
								Rs2ItemModel energyRestoreItem = Rs2Bank.bankItems().stream()
									.filter(rs2Item -> Rs2Potion.getRestoreEnergyPotionsVariants().stream()
										.anyMatch(variant -> rs2Item.getName().toLowerCase().contains(variant.toLowerCase())))
									.min(Comparator.comparingInt(rs2Item -> getDoseFromName(rs2Item.getName())))
									.orElse(null);

								if (energyRestoreItem == null)
								{
									ProjectX.showMessage("Unable to find Restore Energy Potion but hasItem?");
									ProjectX.stopPlugin(plugin);
									return;
								}

								withdrawAndDrink(energyRestoreItem.getName());
							}
							else if (hasStaminaPotion)
							{
								Rs2ItemModel staminaPotionItem = Rs2Bank.bankItems().stream()
									.filter(rs2Item -> rs2Item.getName().toLowerCase().contains(Rs2Potion.getStaminaPotion().toLowerCase()))
									.min(Comparator.comparingInt(rs2Item -> getDoseFromName(rs2Item.getName())))
									.orElse(null);

								if (staminaPotionItem == null)
								{
									ProjectX.showMessage("Unable to find Stamina Potion but hasItem?");
									ProjectX.stopPlugin(plugin);
									return;
								}

								withdrawAndDrink(staminaPotionItem.getName());
							}
							else
							{
								ProjectX.showMessage("Unable to find Stamina Potion OR Energy Restore Potions");
								ProjectX.stopPlugin(plugin);
								return;
							}
						}

						if (Rs2Player.getHealthPercentage() <= config.eatAtPercent())
						{
							while (Rs2Player.getHealthPercentage() < 100 && isRunning())
							{
								if (!Rs2Bank.hasItem(config.food().getId()))
								{
									ProjectX.showMessage("Missing Food in Bank!");
									ProjectX.stopPlugin(plugin);
									break;
								}

								Rs2Bank.withdrawOne(config.food().getId());
								Rs2Inventory.waitForInventoryChanges(1800);
								Rs2Player.useFood();
								sleepUntil(() -> !Rs2Inventory.hasItem(config.food().getId()));
							}

							if (Rs2Inventory.hasItem(ItemID.JUG_EMPTY))
							{
								Rs2Bank.depositAll(ItemID.JUG_EMPTY);
								Rs2Inventory.waitForInventoryChanges(1800);
							}
						}

						int requiredEssence = Rs2Inventory.emptySlotCount() + Rs2Inventory.getRemainingCapacityInPouches();

						if (!Rs2Bank.hasBankItem(config.essence().getItemId(), requiredEssence))
						{
							ProjectX.showMessage("Not enough essence to full run");
							ProjectX.stopPlugin(plugin);
							return;
						}

						if (Rs2Inventory.hasAnyPouch())
						{
							while (!Rs2Inventory.allPouchesFull() && isRunning())
							{
								Rs2Bank.withdrawAll(config.essence().getItemId());
								Rs2Inventory.fillPouches();
								Rs2Inventory.waitForInventoryChanges(1800);
							}
						}

						Rs2Bank.withdrawAll(config.essence().getItemId());
						Rs2Inventory.waitForInventoryChanges(1800);

						Rs2Bank.closeBank();
						sleepUntil(() -> !Rs2Bank.isOpen());
						break;
					case RUNNING_TO_ALTAR:
						if (plugin.isBreakHandlerEnabled())
						{
							BreakHandlerScript.setLockState(true);
						}

						if (config.path().equals(Path.SHORT))
						{
							if (config.directInteract() && ProjectX.isPluginEnabled(GpuPlugin.class))
							{
								var altarModel = ProjectX.getRs2TileObjectCache().query().withId(ObjectID.RC_ZMI_DUNGEON_CRACKED_CENTER_ALTAR).within(Constants.SCENE_SIZE).nearest();
								if (Rs2Camera.getPitch() < 210 || Rs2Camera.getPitch() > 280)
								{
									int randomPitch = Rs2Random.nextInt(220, 260, 1, false);
									Rs2Camera.setPitch(randomPitch);
									sleepUntil(() -> Rs2Camera.getPitch() == randomPitch);
								}
								if (Rs2Camera.getZoom() != 128)
								{
									Rs2Camera.setZoom(128);
									sleepUntil(() -> Rs2Camera.getZoom() == 128);
								}

								if (altarModel != null) altarModel.click("craft-rune");
								sleepUntil(this::isNearAltar, 30000);
							}
							else
							{
								Rs2Walker.walkTo(config.path().getWorldPoint());
							}
						}
						else
						{
							ProjectX.getRs2TileObjectCache().query().withId(ObjectID.RC_ZMI_DUNGEON_WALL_CRACK_ENTRANCE).interact("squeeze-through");
							sleepUntil(this::isNearAltar, 10000);
						}
						break;
				}

				long endTime = System.currentTimeMillis();
				long totalTime = endTime - startTime;
				System.out.println("Total time for loop " + totalTime);

			}
			catch (Exception ex)
			{
				ProjectX.logStackTrace(this.getClass().getSimpleName(), ex);
				ProjectX.log("Error in Ourania Altar Script: " + ex.getMessage());
			}
		}, 0, 1000, TimeUnit.MILLISECONDS);
		return true;
	}

	private boolean hasStateChanged()
	{
		if (ProjectX.isDebug())
		{
			ProjectX.log("State: " + state);
		}
		if (state == null)
		{
			return true;
		}
		if (hasRequiredItems() && !isNearAltar())
		{
			return true;
		}
		if (hasRequiredItems() && isNearAltar())
		{
			return true;
		}
		if ((!hasRequiredItems() && isNearAltar()) || (!hasRequiredItems() && !isNearEniola()))
		{
			return true;
		}
		if (!hasRequiredItems() && isNearEniola())
		{
			return true;
		}
		return false;
	}

	private OuraniaState updateState()
	{
		if (hasRequiredItems() && !isNearAltar())
		{
			return OuraniaState.RUNNING_TO_ALTAR;
		}
		if (hasRequiredItems() && isNearAltar())
		{
			return OuraniaState.CRAFTING;
		}
		if ((!hasRequiredItems() && isNearAltar()) || (!hasRequiredItems() && !isNearEniola()))
		{
			return OuraniaState.RESETTING;
		}
		if (!hasRequiredItems() && isNearEniola())
		{
			return OuraniaState.BANKING;
		}
		return null;
	}

	private boolean hasRequiredItems()
	{
		if (Rs2Inventory.hasAnyPouch())
		{
			boolean pouchesContainEssence = !Rs2Inventory.allPouchesEmpty();
			boolean inventoryContainsEssence = Rs2Inventory.hasItem(config.essence().getItemId());
			return pouchesContainEssence || inventoryContainsEssence;
		}
		else
		{
			return Rs2Inventory.hasItem(config.essence().getItemId());
		}
	}

	private boolean isNearAltar()
	{
		return plugin.getOuraniaAltarArea().contains(Rs2Player.getWorldLocation());
	}

	private boolean isNearEniola()
	{
		Rs2NpcModel eniola = ProjectX.getRs2NpcCache().query().withId(NpcID.RC_ZMI_BANKER).nearest();
		if (eniola == null)
		{
			return false;
		}
		return Rs2Player.getWorldLocation().distanceTo2D(eniola.getWorldLocation()) < 12;
	}

	private void withdrawAndDrink(String potionItemName)
	{
		String simplifiedPotionName = potionItemName.replaceAll("\\s*\\(\\d+\\)", "").trim();
		Rs2Bank.withdrawOne(potionItemName);
		Rs2Inventory.waitForInventoryChanges(1800);
		Rs2Inventory.interact(potionItemName, "drink");
		Rs2Inventory.waitForInventoryChanges(1800);
		if (Rs2Inventory.hasItem(simplifiedPotionName))
		{
			Rs2Bank.depositOne(simplifiedPotionName);
			Rs2Inventory.waitForInventoryChanges(1800);
		}
		if (Rs2Inventory.hasItem(ItemID.VIAL_EMPTY))
		{
			Rs2Bank.depositOne(ItemID.VIAL_EMPTY);
			Rs2Inventory.waitForInventoryChanges(1800);
		}
	}

	private boolean isOnMassWorld()
	{
		return massWorlds.contains(Rs2Player.getWorld());
	}

	private int getDoseFromName(String potionItemName)
	{
		Pattern pattern = Pattern.compile("\\((\\d+)\\)$");
		Matcher matcher = pattern.matcher(potionItemName);
		if (matcher.find())
		{
			return Integer.parseInt(matcher.group(1));
		}
		return 0;
	}
}
